### PR TITLE
Adds Buildkite integration & removes CircleCI

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR adds support for building the library in Buildkite instead of CircleCI. It also improves a few other things that we have already applied to other libraries:

* 5be3a8fad2649db1650df7fdfbbd9f150ecbc774 adds plugin DSL support and updates the publish-to-s3 plugin
* af702cad19f61a66887d45d22ff24c99cba55008 stops tracking the `.idea/compiler.xml` file. We have made this change a long time ago for other projects such as this one for WordPress-Android: https://github.com/wordpress-mobile/WordPress-Android/pull/13755
* During 5be3a8fad2649db1650df7fdfbbd9f150ecbc774 I've removed the deprecated `kotlin-android-extensions` plugin. I've considered replacing it with the [newer viewBinding feature](https://developer.android.com/topic/libraries/view-binding), however there were only a few usages to be updated, so I just replaced these with the regular `findViewById` method in cd445db670ea8b1f65d4550dc57c168a89ce63be. This is consistent with the rest of the project and admittedly since I've never worked with `viewBinding` feature, I didn't feel like spending extra time learning it for this.
* I've removed the `kotlin-parcelize` plugin in 62c08a1674c93989f0b4e4a3be476766afd2aa37. I thought this was going to be necessary when I worked on `5be3a8fad2649db1650df7fdfbbd9f150ecbc774` which turned out to be incorrect.

### CODE REVIEW

**Review by renanferrari — APPROVED** [review]
Hey @oguzkocer 👋 Thanks for the changes!

I've reviewed the changes in 5be3a8fad2649db1650df7fdfbbd9f150ecbc774 specifically, as you requested, and everything looks good to me 👍 

I have also tested the app and can confirm everything is working as expected around that screen.

LGTM 👍

**Comment by oguzkocer** [comment]
Thank you for the review @renanferrari 🙇

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:19]: 🤔 💭  I notice that we had a `restore_gradle_cache` in the CircleCI config while we don't have a similar one in this Buildkite config.

This can obviously be added later and is not a blocker to this PR, but it could be nice to use [our `save_cache` / `restore_cache` actions](https://github.com/Automattic/bash-cache-buildkite-plugin) to replicate that gradle caching here. Except if not replicating this here was intentional because you think it's not worth it (or that caching + restoring takes more time that it saves)?

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:29]: 👍  for _not_ including the `--stacktrace` that we had in CircleCI before. Less verbose logs are easier to read, and having stacktrace by default on _every_ gradle commands is seldom useful, so I like these more compact logs.

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:34]: 👍  btw, I noticed that the [CircleCI config made the publication job only depend on test but _not_ on lint](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/74/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47L53-L64), and I'm not sure why we allowed publication if the lint failed back then… but good choice to depend on both now! 👌

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:2]: 💄  Nit: since this Docker image is not only used for the publish step, but also the lint and test steps, not sure why the anchor name was chosen to be _this_ specific 🤷 

May I suggest simply using `&docker-container` for this instead?

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:41]: Note to self (and any future reviewer):

 - I notice that CircleCI config had the `filters: { tags: { only: /.*/ } }` so at first I incorrectly thought that this meant that the publish job was _only_ run on tags, but not on regular pushes on branches. Which would mean we'd have a difference in this Buildkite config compared to that old setup.
 - But my assumption was wrong, because in fact [CircleCI will _always_ build on all branches (unless you specify an explicit `filters: { branches: … }`)](https://circleci.com/docs/2.0/configuration-reference/#branches) so, contradictory to intuition, that syntax in the old CircleCI config meant "also build on tags only matching `.*` (aka all tags)… [in addition to building for all branches like it will do by default"

So TL;DR: that confirms that despite the apparences, the new Buildkite config for this step _does_ match the behavior we had in the old CircleCI config for that job 👌

**Review by AliSoftware — APPROVED** [review]
LGTM! 👍 

I mostly reviewed the CI config files, as well as the `build.gradle` changes (as far as my understanding of gradle allowed me to judge it 😅 ), and didn't review the changes in Kotlin source files though, but seeing that @renanferrari already reviewed those changes I trust this part is ok 🙂 

💄 I left just a nitpick on the name of the &anchor for the Docker plugin in the Buildkite config, but it's really a minutiae.
🤔  Also left a question about gradle cache (not a blocker either)

Other that this, it looks good to go 👍 👏 

---

PS: Remember to disable the CircleCI integration for this repo/project once this PR lands, so that it's not run as part of PR checks anymore 🙂

File: .buildkite/pipeline.yml
Comment (oguzkocer) [reply .buildkite/pipeline.yml:19]: This is a project planned to be undertaken in the future, once we have the library versioning improvements in place. We have _some_ caching already in place (such as the Gradle wrapper which needs to be updated in the Docker image 🤔 ), but not to an extent of what we did in CircleCI.

Having said that, the way we cached and restored the Gradle cache in CircleCI was not very efficient and sometimes costed us more time to restore the cache than a simple run. We also kept building on top of the existing cache and never purged anything which made the cache quite large. This didn't matter for these smaller libraries, but costed 4-5 minutes for bigger projects such as WordPress-Android if I remember correctly.

We are considering of taking a different approach to caching with Buildkite when the time comes. There will certainly be an RFC to discuss it, so I'll leave it at that for now.

Comment (AliSoftware) [reply .buildkite/pipeline.yml:19]: Makes perfect sense, thanks for the detailed rationale 👍

File: .buildkite/pipeline.yml
Comment (oguzkocer) [reply .buildkite/pipeline.yml:2]: I keep going back and forth on that decision, but the main reason this specific name is picked (besides being consistent with other projects) is that the AWS credentials are passed through to the Docker container. Although it's very unlikely that we'd accidentally leak them, I felt that by choosing a specific name and having the comment above the credentials may lead us (mostly myself) to be more cautious.

If that doesn't sound like a reasonable precaution, I don't mind renaming it to `docker-container`. As I said, I have been feeling unsure about this each time I replicated it in another project, so it'd be very easy to change my mind :)

Comment (AliSoftware) [reply .buildkite/pipeline.yml:2]: Ah I see, I didn't realize that was the intention behind the naming. Personally that name didn't really make me think "I'm announcing that S3 credentials are in this image" so that wouldn't have made the expected effect on me tbh 🙃 

If you want to highlight that aspect, maybe `docker-container-with-s3-tools` or `docker-container-with-s3-access` would make it clearer?

But again, it's really just a minutiae and probably not worth debating over much; I'd say just go with your guts and choose what makes sense for you, and don't let this block this PR from being merged 🙂

Comment (oguzkocer) [reply .buildkite/pipeline.yml:2]: Since this is the name we use in other projects, let's keep it as is then. We can re-visit them all together in the future. _And that's not because I am on a different computer and can't make changes right now 🤫_

**Pull request merged**
